File: 13_Roman_To_Integer.js
function romanToInt(s){
    const romanDict = {
        'I': 1,
        'V': 5,
        'X': 10,
        'L': 50,
        'C': 100,
        'D': 500,
        'M': 1000
      }
      let total = 0 , prev = 0 
//       Start from the end of the string 
      for(let i = s.length-1 ; i >= 0 ; i--){
//         access time for dictionary is O(1) so we are looking into the dictionary with the value at the i'th position 
        const curr = romanDict[s.charAt(i)]
        if(curr < prev){
            total = total - curr
        }else{
            total = total + curr
        }
//         initialize prev with curr value to use as prev for the next loop 
        prev = curr 
      }
      return total
}

s = "MCMXCIV"
console.log(romanToInt(s))

File: 13_Roman_to_Integer.java
class Solution{
    public int romanToInt(String s){
        HashMap<Character , Integer> romanDict = new HashMap<>();       
        romanDict.put('I',1);
        romanDict.put('V',5);
        romanDict.put('X',10);
        romanDict.put('L',50);
        romanDict.put('C',100);
        romanDict.put('D',500);
        romanDict.put('M',1000);
        int total = 0 , prev = 0;
        for(int i = s.length()-1 ; i >= 0 ; i--){
           int curr = romanDict.get(s.charAt(i));
           if(curr < prev){
              total = total - curr;
           }else{
              total = total + curr;
           }
           prev = curr;
        }
        return total;
    }
}